core/index/segment_reader_impl.cpp

```diff
@@ -0,0 +1,346 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2022 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrey Abramov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "segment_reader_impl.hpp"
+
+#include <vector>
+
+#include "analysis/token_attributes.hpp"
+#include "index/index_meta.hpp"
+#include "utils/index_utils.hpp"
+#include "utils/type_limits.hpp"
+
+#include <absl/strings/str_cat.h>
+
+namespace irs {
+namespace {
+
+class AllIterator final : public doc_iterator {
+ public:
+  explicit AllIterator(doc_id_t docs_count) noexcept
+    : max_doc_{doc_limits::min() + docs_count - 1} {}
+
+  bool next() noexcept override {
+    if (doc_.value < max_doc_) {
+      ++doc_.value;
+      return true;
+    } else {
+      doc_.value = doc_limits::eof();
+      return false;
+    }
+  }
+
+  doc_id_t seek(doc_id_t target) noexcept override {
+    doc_.value = target <= max_doc_ ? target : doc_limits::eof();
+
+    return doc_.value;
+  }
+
+  doc_id_t value() const noexcept override { return doc_.value; }
+
+  attribute* get_mutable(irs::type_info::type_id type) noexcept override {
+    return irs::type<document>::id() == type ? &doc_ : nullptr;
+  }
+
+ private:
+  document doc_;
+  doc_id_t max_doc_;  // largest valid doc_id
+};
+
+class MaskDocIterator final : public doc_iterator {
+ public:
+  MaskDocIterator(doc_iterator::ptr&& it, const document_mask& mask) noexcept
+    : mask_{mask}, it_{std::move(it)} {}
+
+  bool next() override {
```

Comment:
please write final for final methods it's not really difficult but really helpful if sometime class became not final